File: 2nd Year/2nd Semester/MPP/PROIECT-C#/Lab1/TriatlonNetworking/dto/DTOUtils.cs
using System;
using System.Collections.Generic;
using System.Linq;
using TriatlonModel;

namespace TriatlonNetworking.dto
{
    public static class DTOUtils
    {
        // Convert domain objects to DTOs

        public static ArbitruDTO GetDTO(Arbitru arbitru)
        {
            if (arbitru == null) return null;
            return new ArbitruDTO(
                arbitru.Id,
                arbitru.Username,
                arbitru.Password,
                arbitru.FirstName,
                arbitru.LastName
            );
        }

        public static ParticipantDTO GetDTO(Participant participant)
        {
            if (participant == null) return null;
            var dto = new ParticipantDTO(
                participant.Id,
                participant.FirstName,
                participant.LastName
            );

            // Copy punctaje from participant to DTO
            foreach (TipProba tipProba in Enum.GetValues(typeof(TipProba)))
            {
                int punctaj = participant.GetPunctajProba(tipProba);
                dto.SetPunctajProba(GetTipProbaDTO(tipProba), punctaj);
            }

            return dto;
        }

        public static RezultatDTO GetDTO(Rezultat rezultat)
        {
            if (rezultat == null) return null;
            return new RezultatDTO(
                rezultat.Id,
                GetDTO(rezultat.Participant),
                GetDTO(rezultat.Arbitru),
                GetTipProbaDTO(rezultat.TipProba),
                rezultat.Points
            );
        }

        public static ProbaDTO GetDTO(int? id, TipProba tipProba, Arbitru arbitru)
        {
            if (tipProba == null) return null;
            return new ProbaDTO(
                id,
                GetTipProbaDTO(tipProba),
                GetDTO(arbitru)
            );
        }

        public static TipProbaDTO GetTipProbaDTO(TipProba tipProba)
        {
            return tipProba switch
            {
                TipProba.NATATIE => TipProbaDTO.NATATIE,
                TipProba.CICLISM => TipProbaDTO.CICLISM,
                TipProba.ALERGARE => TipProbaDTO.ALERGARE,
                _ => throw new ArgumentException("Invalid TipProba value")
            };
        }

        // Convert arrays of domain objects to arrays of DTOs

        public static ArbitruDTO[] GetDTO(Arbitru[] arbitri)
        {
            if (arbitri == null) return null;
            return arbitri.Select(GetDTO).ToArray();
        }

        public static RezultatDTO[] GetDTO(Rezultat[] rezultate)
        {
            if (rezultate == null) return null;
            return rezultate.Select(GetDTO).ToArray();
        }

        // Convert DTOs to domain objects

        public static Arbitru GetFromDTO(ArbitruDTO dto)
        {
            if (dto == null) return null;
            return new Arbitru(
                dto.Id ?? 0,
                dto.Username,
                dto.Password,
                dto.FirstName,
                dto.LastName
            );
        }

        public static Participant GetFromDTO(ParticipantDTO dto)
        {
            if (dto == null) return null;
            var participant = new Participant(
                dto.Id ?? 0,
                dto.FirstName,
                dto.LastName
            );

            // Copy punctaje from DTO to participant
            foreach (TipProbaDTO tipProbaDTO in Enum.GetValues(typeof(TipProbaDTO)))
            {
                TipProba tipProba = GetFromDTO(tipProbaDTO);
                if (tipProba != null)
                {
                    int punctaj = dto.GetPunctajProba(tipProbaDTO);
                    participant.SetPunctajProba(tipProba, punctaj);
                }
            }

            return participant;
        }

        public static Rezultat GetFromDTO(RezultatDTO dto)
        {
            if (dto == null) return null;
            return new Rezultat(
                dto.Id ?? 0,
                GetFromDTO(dto.Participant),
                GetFromDTO(dto.Arbitru),
                GetFromDTO(dto.TipProba),
                dto.Punctaj
            );
        }

        public static TipProba GetFromDTO(TipProbaDTO tipProbaDTO)
        {
            return tipProbaDTO switch
            {
                TipProbaDTO.NATATIE => TipProba.NATATIE,
                TipProbaDTO.CICLISM => TipProba.CICLISM,
                TipProbaDTO.ALERGARE => TipProba.ALERGARE,
                _ => throw new ArgumentException("Invalid TipProbaDTO value")
            };
        }

        // Convert arrays of DTOs to arrays of domain objects

        public static Arbitru[] GetFromDTO(ArbitruDTO[] dtos)
        {
            if (dtos == null) return null;
            return dtos.Select(GetFromDTO).ToArray();
        }

        public static Rezultat[] GetFromDTO(RezultatDTO[] dtos)
        {
            if (dtos == null) return null;
            return dtos.Select(GetFromDTO).ToArray();
        }

        public static IEnumerable<Participant> GetFromDTO(ParticipantDTO[] participantiDTO)
        {
            if (participantiDTO == null) return null;
            var participanti = new Dictionary<int?, Participant>();
            foreach (var dto in participantiDTO)
            {
                if (dto.Id.HasValue)
                    participanti[dto.Id] = GetFromDTO(dto);
            }

            return participanti.Values;
        }
    }
}

File: 2nd Year/2nd Semester/MPP/Triatlon-Java/TriatlonNetworking/src/main/java/triatlon/network/dto/DTOUtils.java
package triatlon.network.dto;

import triatlon.model.Arbitru;
import triatlon.model.Participant;
import triatlon.model.Rezultat;
import triatlon.model.TipProba;

import java.util.HashMap;
import java.util.Map;

public class DTOUtils {

    // Convert domain objects to DTOs

    public static ArbitruDTO getDTO(Arbitru arbitru) {
        if (arbitru == null) return null;
        return new ArbitruDTO(
                arbitru.getId(),
                arbitru.getUsername(),
                arbitru.getPassword(),
                arbitru.getFirst_name(),
                arbitru.getLast_name()
        );
    }

    public static ParticipantDTO getDTO(Participant participant) {
        if (participant == null) return null;
        ParticipantDTO dto = new ParticipantDTO(
                participant.getId(),
                participant.getFirst_name(),
                participant.getLast_name()
        );

        // Copy punctaje from participant to DTO
        for (TipProba tipProba : TipProba.values()) {
            int punctaj = participant.getPunctajProba(tipProba);
            dto.setPunctajProba(getTipProbaDTO(tipProba), punctaj);
        }

        return dto;
    }

    public static RezultatDTO getDTO(Rezultat rezultat) {
        if (rezultat == null) return null;
        return new RezultatDTO(
                rezultat.getId(),
                getDTO(rezultat.getParticipant()),
                getDTO(rezultat.getArbitru()),
                getTipProbaDTO(rezultat.getTipProba()),
                rezultat.getPunctaj()
        );
    }

    public static ProbaDTO getDTO(Integer id, TipProba tipProba, Arbitru arbitru) {
        if (tipProba == null) return null;
        return new ProbaDTO(
                id,
                getTipProbaDTO(tipProba),
                getDTO(arbitru)
        );
    }

    public static TipProbaDTO getTipProbaDTO(TipProba tipProba) {
        if (tipProba == null) return null;
        switch (tipProba) {
            case NATATIE: return TipProbaDTO.NATATIE;
            case CICLISM: return TipProbaDTO.CICLISM;
            case ALERGARE: return TipProbaDTO.ALERGARE;
            default: return null;
        }
    }

    // Convert arrays of domain objects to arrays of DTOs

    public static ArbitruDTO[] getDTO(Arbitru[] arbitri) {
        if (arbitri == null) return null;
        ArbitruDTO[] dtos = new ArbitruDTO[arbitri.length];
        for (int i = 0; i < arbitri.length; i++) {
            dtos[i] = getDTO(arbitri[i]);
        }
        return dtos;
    }

    public static RezultatDTO[] getDTO(Rezultat[] rezultate) {
        if (rezultate == null) return null;
        RezultatDTO[] dtos = new RezultatDTO[rezultate.length];
        for (int i = 0; i < rezultate.length; i++) {
            dtos[i] = getDTO(rezultate[i]);
        }
        return dtos;
    }

    // Convert DTOs to domain objects

    public static Arbitru getFromDTO(ArbitruDTO dto) {
        if (dto == null) return null;
        return new Arbitru(
                dto.getId(),
                dto.getUsername(),
                dto.getPassword(),
                dto.getFirst_name(),
                dto.getLast_name()
        );
    }

    public static Participant getFromDTO(ParticipantDTO dto) {
        if (dto == null) return null;
        Participant participant = new Participant(
                dto.getId(),
                dto.getFirst_name(),
                dto.getLast_name()
        );

        // Copy punctaje from DTO to participant
        for (TipProbaDTO tipProbaDTO : TipProbaDTO.values()) {
            TipProba tipProba = getFromDTO(tipProbaDTO);
            if (tipProba != null) {
                int punctaj = dto.getPunctajProba(tipProbaDTO);
                participant.setPunctajProba(tipProba, punctaj);
            }
        }

        return participant;
    }

    public static Rezultat getFromDTO(RezultatDTO dto) {
        if (dto == null) return null;
        return new Rezultat(
                dto.getId(),
                getFromDTO(dto.getParticipant()),
                getFromDTO(dto.getArbitru()),
                getFromDTO(dto.getTipProba()),
                dto.getPunctaj()
        );
    }

    public static TipProba getFromDTO(TipProbaDTO tipProbaDTO) {
        if (tipProbaDTO == null) return null;
        switch (tipProbaDTO) {
            case NATATIE: return TipProba.NATATIE;
            case CICLISM: return TipProba.CICLISM;
            case ALERGARE: return TipProba.ALERGARE;
            default: return null;
        }
    }

    // Convert arrays of DTOs to arrays of domain objects

    public static Arbitru[] getFromDTO(ArbitruDTO[] dtos) {
        if (dtos == null) return null;
        Arbitru[] arbitri = new Arbitru[dtos.length];
        for (int i = 0; i < dtos.length; i++) {
            arbitri[i] = getFromDTO(dtos[i]);
        }
        return arbitri;
    }

    public static Rezultat[] getFromDTO(RezultatDTO[] dtos) {
        if (dtos == null) return null;
        Rezultat[] rezultate = new Rezultat[dtos.length];
        for (int i = 0; i < dtos.length; i++) {
            rezultate[i] = getFromDTO(dtos[i]);
        }
        return rezultate;
    }

    public static Object getFromDTO(ParticipantDTO[] participantiDTO) {
        if (participantiDTO == null) return null;
        Map<Integer, Participant> participanti = new HashMap<>();
        for (ParticipantDTO dto : participantiDTO) {
            participanti.put(dto.getId(), getFromDTO(dto));
        }
        return participanti.values();
    }
}